feature/security/gnsi/certz/tests/internal/setup_service/setup_service.go

```diff
@@ -0,0 +1,606 @@
+			break
+		}
+		time.Sleep(10 * time.Second)
+		if i == retries-1 {
+			t.Fatalf("gNSI service RPC  did not succeed ~%d*10s after rotate. Certz/Rotate failed. FinalizeRequest will not be sent", retries)
+		}
+	}
+	finalizeRequest := &certzpb.RotateCertificateRequest_FinalizeRotation{FinalizeRotation: &certzpb.FinalizeRequest{}}
+	rotateCertRequest = &certzpb.RotateCertificateRequest{
+		ForceOverwrite: false,
+		SslProfileId:   profileID,
+		RotateRequest:  finalizeRequest,
+	}
+	if err := rotateRequestClient.Send(rotateCertRequest); err != nil {
+		t.Fatalf("Error sending rotate finalize request: %v", err)
+	}
+	if err := rotateRequestClient.CloseSend(); err != nil {
+		t.Fatalf("Error sending rotate close send request: %v", err)
+	}
+	return true
+}
+
+// CertGeneration function to create test data for use in TLS tests.
+func CertGeneration(t *testing.T, dirPath string) error {
+	cmd := exec.Cmd{
+		Path:   "./mk_cas.sh",
+		Stdout: os.Stdout,
+		Stderr: os.Stderr,
+	}
+	cmd.Dir = dirPath
+	t.Logf("Executing cert generation command %v.", cmd)
+	err := cmd.Start()
+	if err != nil {
+		t.Fatalf("Cert generation command failed with error:%v.", err)
+	}
+	err = cmd.Wait()
+	if err != nil {
+		t.Fatalf("Failed to run cert generation command during wait with error:%v.", err)
+	}
+	return err
+}
+
+// CertCleanup function to clean out the certificate content under test_data.
+func CertCleanup(t *testing.T, dirPath string) error {
+	cmd := exec.Command("./cleanup.sh")
+	cmd.Dir = dirPath
+	t.Logf("Executing cleanup command")
+	if err := cmd.Run(); err != nil {
+		t.Fatalf("Testdata cleanup command failed with error:%v.", err)
+	}
+	return nil
+}
+
+// ReadDecodeServerCertificate function to read and decode server certificates to extract the SubjectAltName and validate.
+func ReadDecodeServerCertificate(t *testing.T, serverCertzFile string) (san string) {
+	sc, err := os.ReadFile(serverCertzFile)
+	if err != nil {
+		t.Fatalf("Failed to read certificate with error: %v.", err)
+	}
+	block, _ := pem.Decode(sc)
+	if block == nil {
+		t.Fatalf("Failed to parse PEM block containing the public key.")
+	}
+	sCert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		t.Fatalf("Failed to parse certificate with error: %v.", err)
+	}
+	san = sCert.DNSNames[0]
+	t.Logf("ServerAltName:%s.", san)
+	if sn != san {
+		t.Fatalf("ServerAltName validation failed for %s.", serverCertzFile)
+	}
+	return san
+}
+
+// VerifyGnsi function to validate the gNSI service RPC after successful rotation.
+func VerifyGnsi(t *testing.T, caCert *x509.CertPool, san, serverAddr, username, password string, cert tls.Certificate) bool {
+	credOpts := []grpc.DialOption{grpc.WithTransportCredentials(credentials.NewTLS(
+		&tls.Config{
+			Certificates: []tls.Certificate{cert},
+			RootCAs:      caCert,
+			ServerName:   san,
+		}))}
+	creds := &rpcCredentials{&creds.UserPass{Username: username, Password: password}}
```

Comment:
shouldn't the authentication here be with the certs and not a username/password?